firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java

```diff
@@ -4578,22 +4578,324 @@ public void testGetWaitsForConnection() throws DatabaseException, InterruptedExc
+        };
+
+    try {
+      await(childWriteNode.setValue(val));
+      new ReadFuture(
+              childReadNode,
+              events -> {
+                assertEquals(1, events.size());
+                assertEquals(val, events.get(0).getSnapshot().getValue());
+                return true;
+              })
+          .timedGet();
+      childReadNode.addValueEventListener(listener);
+      DatabaseReference parentReadNode = readDb.getReference().child(parentNodeKey);
+      IntegrationTestHelpers.waitFor(semaphore);
+      DataSnapshot snapshot = await(parentReadNode.get());
+      assertEquals(snapshot.child(childNodeKey).getValue(), val);
+      await(parentWriteNode.setValue(new MapBuilder().put(childNodeKey, val).build()));
+      DataSnapshot parentSnapshot = await(childReadNode.get());
+      assertEquals(parentSnapshot.getValue(), val);
+      childReadNode.removeEventListener(listener);
+    } catch (ExecutionException | TimeoutException | TestFailure e) {
+      fail("get threw an exception: " + e);
+    }
+  }
+
+  @Test
+  public void testLimitToGetWithListenerOnlyFiresOnce()
+      throws DatabaseException, InterruptedException, ExecutionException {
+    FirebaseApp readApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseApp writeApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase readDb = FirebaseDatabase.getInstance(readApp);
+    FirebaseDatabase writeDb = FirebaseDatabase.getInstance(writeApp);
+    String refKey = UUID.randomUUID().toString();
+    DatabaseReference writeRef = writeDb.getReference().child(refKey);
+    DatabaseReference readRef = readDb.getReference().child(refKey);
+    Map<String, Object> expected =
+        new MapBuilder().put("child1", "test1").put("child2", "test2").build();
+    try {
+      await(writeRef.setValue(expected));
+      ReadFuture.untilEquals(readRef, expected).timedGet();
+    } catch (Exception e) {
+      fail("Test failed with exception" + e);
+    }
+
+    Query query = readRef.limitToFirst(1);
+    DataSnapshot snapshot = await(query.get());
+    Map<String, Object> map = new MapBuilder().put("child1", "test1").build();
+    assertEquals(map, snapshot.getValue());
+  }
+
+  @Test
+  public void testDisjointedListenAndGet() {
+    FirebaseApp readApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseApp writeApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase readDb = FirebaseDatabase.getInstance(readApp);
+    FirebaseDatabase writeDb = FirebaseDatabase.getInstance(writeApp);
+    String parentKey = UUID.randomUUID().toString();
+    DatabaseReference writeRef = writeDb.getReference().child(parentKey);
+    DatabaseReference child1Ref = readDb.getReference().child(parentKey).child("child1");
```

Comment:
Feel free to do this in a separate PR, we could definitely abstract some of this setup into a helper, potentially even a some sort of TestFixture class.